### PR TITLE
[BEAM-9037] Instant and duration as logical type

### PR DESCRIPTION
**Pull request by alexvanboxel**
The proto schema includes Timestamp and Duration with nano precision.
The logical types should be promoted to the core logical types, so they
can be handled on various IO's as standard mandatory conversions.

This means that the logical type should use the proto specific Timestamp
and Duration but the java 8 Instant and Duration.

------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by alexvanboxel** [comment]
@TheNeuralBit @reuvenlax this PR promotes the Timestamp+Duration to NanosInstant and NanosDuration. Please see the PR for more information and reference to the design document.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/logicaltypes/NanosType.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+/** Base class for types representing timestamps or durations as nanoseconds. */
+abstract class NanosType<T> implements Schema.LogicalType<T, Row> {
+  private final String identifier;
+
+  protected static final Schema SCHEMA;
+
+  static {
+    Schema schema = Schema.builder().addInt64Field("seconds").addInt32Field("nanos").build();
+    schema.setUUID(UUID.fromString("dc5915f1-6db8-3134-9444-4980fb088e92"));
+    SCHEMA = schema;
```
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/logicaltypes/NanosType.java]: this is a bit ugly - should we instead just make it an instance variable?

**Comment by TheNeuralBit** [comment]
@reuvenlax FYI the static schema stuff is a backport of #10569 - we were discussing alternatives over there.

**Comment by alexvanboxel** [comment]
> @reuvenlax FYI the static schema stuff is a backport of #10569 - we were discussing alternatives over there.

Let's first agree on: is replacing the proto specific into these standard logical types. Then I want to put in the work of fixing this. I think we indeed need instance variables. But the ByteBuffy coding was using the static SCHEMA as well... so it's not that easy. So question 1) is this a good idea:  Instant and duration as logical type?

**Comment by TheNeuralBit** [comment]
I'm +1 for instant and duration as logical type. I think we should do the same for millis_instant, but that will require a lot of other changes.

**Comment by alexvanboxel** [comment]
Run Java PreCommit

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/logicaltypes/NanosType.java (hunk 1 above)
Comment (alexvanboxel) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/logicaltypes/NanosType.java]: Switched to instance variables. Proto conversion used by ByteBuddy still uses a static version of the schema for performance reasons.

**Comment by TheNeuralBit** [comment]
@alexvanboxel I can go ahead and merge if you just add a `v1` to the URNs

**Comment by alexvanboxel** [comment]
> @alexvanboxel I can go ahead and merge if you just add a `v1` to the URNs

Done

**Comment by TheNeuralBit** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by alexvanboxel** [comment]
I've rebased and squashed. Hopefully this avoid the Jenkins trouble... Fingers crossed.

**Comment by alexvanboxel** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by alexvanboxel** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by alexvanboxel** [comment]
I shouldn't have triggered Run Java_Examples_Dataflow PreCommit

it was green. Grumple.

**Comment by alexvanboxel** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by alexvanboxel** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by alexvanboxel** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by alexvanboxel** [comment]
@TheNeuralBit all green, v1 added, rebased against master and squached. Ready to merge.

**Comment by alexvanboxel** [comment]
@TheNeuralBit or @reuvenlax as this is approved can some this be merged. I can't do it as I'm only a humble contributor. I like to rebase my other work then on master. Thanks.

**Comment by alexvanboxel** [comment]
As promised, schema aware topics adding: @iemejia @aromanenko-dev (this one can be merged)

**Comment by TheNeuralBit** [comment]
Sorry for the delay, I was out of the office for a few days. Merging now!

**Comment by alexvanboxel** [comment]
Thank you! Happy Happy :-)

**Pull request merged**